File: bnovate/bnovate/report/revenue_analytics/revenue_analytics.js
// Copyright (c) 2016, libracore and contributors
// For license information, please see license.txt
/* eslint-disable */
frappe.require("/assets/bnovate/js/modals.js")  // provides bnovate.modals

frappe.query_reports["Revenue Analytics"] = {
	filters: [
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": new Date(new Date().getFullYear(), 0, 1), // First day of calendar year.
			"reqd": 1
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": new Date(new Date().getFullYear(), 11, 31), // Last day of calendar year.
			"reqd": 1
		},
		{
			"fieldname": "include",
			"label": __("Include"),
			"fieldtype": "Select",
			"options": ["All", "Billed", "Unbilled"],
			"default": "All"
		},
		// {
		// 	"fieldname": "interval",
		// 	"label": __("Time Interval"),
		// 	"fieldtype": "Select",
		// 	"options": ["Month", "Quarter", "Year"],
		// 	"default": "Month"
		// },
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
	],
	tree: true,
	name_field: "revenue_stream_name",
	parent_field: "parent_revenue_stream",
	initial_depth: 2,
	onload: function (report) {
		this.report = report;
		this.colours = ["darker", "dark", "light"];
		this.max_indent = null;

		bnovate.modals.attach_report_modal("masterModal");
	},
	formatter: function (value, row, col, data, default_formatter) {

		if (this.max_indent === null) {
			this.max_indent = Math.max(...frappe.query_report.data.map(d => d.indent || 0));
		}

		if (data.indent < this.max_indent) {
			return `<span class="coloured ${this.colours[data.indent % this.colours.length]}">${default_formatter(value, row, col, data)}</span>`;
		} else {
			return this.build_master_link(default_formatter(value, row, col, data), data.revenue_stream_name, col.fieldname);
		}

		return default_formatter(value, row, col, data);
	},
	build_master_link(display_text, revenue_stream, month) {
		let from_date = frappe.query_report.filters?.find(f => f.fieldname === "from_date")?.get_value();
		let to_date = frappe.query_report.filters?.find(f => f.fieldname === "to_date")?.get_value();

		if (month !== 'total') {
			const m = moment(month, "YYYY-MM");
			from_date = m.startOf("month").format("YYYY-MM-DD");
			to_date = m.endOf("month").format("YYYY-MM-DD");
		};

		const filters = {
			revenue_stream,
			from_date,
			to_date,
			company: frappe.query_report.filters?.find(f => f.fieldname === "company")?.get_value() || frappe.defaults.get_user_default("Company"),
			include: frappe.query_report.filters?.find(f => f.fieldname === "include")?.get_value() || "All",
		};

		return bnovate.modals.report_link(
			display_text,
			"masterModal",
			"Revenue Analytics Master",
			`Revenue Analytics - ${revenue_stream} - ${month}`,
			filters
		)
	},
};

File: bnovate/bnovate/report/revenue_analytics/revenue_analytics.py
# Copyright (c) 2013, libracore and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.utils import flt, getdate, add_months, add_years
from erpnext import get_company_currency

from bnovate.bnovate.report.revenue_analytics_master.revenue_analytics_master import get_data as get_master_data , validate_filters


def execute(filters=None):
    if not filters:
        filters = {}

    validate_filters(filters)

    columns = get_columns(filters)
    data = get_data(filters)

    return columns, data

def get_columns(filters):
    columns = [
        {
            "fieldname": "revenue_stream_name",
            "label": _("Revenue Stream"),
            "fieldtype": "Data",
            "width": 200
        }
    ]

    from_date = getdate(filters.get("from_date"))
    to_date = getdate(filters.get("to_date"))


    current = from_date
    while current <= to_date:
        month_year = current.strftime("%b %Y")
        fieldname = current.strftime("%Y-%m")
        columns.append({
            "fieldname": fieldname,
            "label": month_year,
            "fieldtype": "Currency",
            "options": "currency",
            "width": 120
        })

        current = add_months(current, 1)

    columns.append({
        "fieldname": "total",
        "label": _("Total"),
        "fieldtype": "Currency",
        "options": "currency",
        "width": 120
    })

    return columns

def get_data(filters):


    # Get all revenue streams in tree order
    revenue_streams = frappe.db.sql("""
        SELECT 
            rs.name, 
            rs.revenue_stream_name, 
            rs.lft, 
            rs.rgt, 
            rs.is_group, 
            rs.parent_revenue_stream,
            COUNT(rs_parents.name) as depth
        FROM `tabRevenue Stream` rs
        LEFT OUTER JOIN `tabRevenue Stream` rs_parents ON rs_parents.lft < rs.lft AND rs_parents.rgt > rs.rgt
        GROUP BY rs.name
        ORDER BY rs.lft
    """, as_dict=True)

    aggregated_data = get_master_data(filters)

    # Aggregate 
    amounts = {}
    for row in aggregated_data:
        key = (row.revenue_stream_name, row.get("month"))
        if key not in amounts:
            amounts[key] = 0

        amounts[key] += row.amount
            

    # Build the data rows with initial values
    data = []
    rows_by_name = {}
    from_date = getdate(filters.get("from_date"))
    to_date = getdate(filters.get("to_date"))

    for rs in revenue_streams:
        row = {
            "name": rs.name,
            "revenue_stream_name": rs.revenue_stream_name,
            "parent_revenue_stream": rs.parent_revenue_stream,
            "indent": rs.depth
        }

        total = 0
        current = from_date
        while current <= to_date:
            month_key = current.strftime("%Y-%m")
            amount = amounts.get((rs.name, month_key), 0)
            row[month_key] = amount
            total += amount
            current = add_months(current, 1)

        row["total"] = total
        data.append(row)
        rows_by_name[rs.name] = row

    # Rollup child amounts into parent revenue streams
    for rs in reversed(revenue_streams):
        if not rs.parent_revenue_stream:
            continue
        child_row = rows_by_name.get(rs.name)
        parent_row = rows_by_name.get(rs.parent_revenue_stream)
        if not child_row or not parent_row:
            continue

        current = from_date
        while current <= to_date:
            month_key = current.strftime("%Y-%m")
            parent_row[month_key] = parent_row.get(month_key, 0) + child_row.get(month_key, 0)
            current = add_months(current, 1)

        parent_row["total"] = parent_row.get("total", 0) + child_row.get("total", 0)

    return data
